### ISSUE TITLE
segfault: tibble expect_equal or expect_identical handling list columns

### ISSUE THREAD

**Issue by colearendt** [r-lib/testthat#985]
It seems like this was a feature added recently? Or perhaps somewhere upstream? In any case, tibble comparison have begun breaking via `expect_equal` and `expect_identical` in some non-CRAN tests that have list columns in them:

```
Error: Can't join on 'git' x 'git' because of incompatible types (list / list)
```

...

And then while trying to create a reprex, I stopped being able to reproduce this... (I suspect I had an old version of a package loaded, perhaps, and needed to restart my R session).

This reprex is pretty close to what I am doing:

``` r
t1 <- tibble::tibble(obj = list(list())); 
testthat::expect_equal(t1, t1); 
testthat::expect_equal(t1[0,], t1[0,])
```

<sup>Created on 2020-02-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

And then I ran `devtools::test()` again and got a segfault in R (on Mac, using R 3.4.3). This has become reproducible now:

```
Traceback:
 1: .Call(`_dplyr_equal_data_frame`, x, y, ignore_col_order, ignore_row_order,     convert)
 2: equal_data_frame(target, current, ignore_col_order = ignore_col_order,     ignore_row_order = ignore_row_order, convert = convert)
 3: all.equal.tbl_df(x, y, ...)
 4: all.equal(x, y, ...)
 5: compare.default(act$val, exp$val)
 6: compare(act$val, exp$val)
 7: expect_identical(vctrs::vec_ptype(content_list), vctrs::vec_ptype(connectapi_ptypes$content))
 8: eval(code, test_env)
 9: eval(code, test_env)
10: withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error)
11: doTryCatch(return(expr), name, parentenv, handler)
12: tryCatchOne(expr, names, parentenv, handlers[[1L]])
13: tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
14: doTryCatch(return(expr), name, parentenv, handler)
15: tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]),     names[nh], parentenv, handlers[[nh]])
16: tryCatchList(expr, classes, parentenv, handlers)
17: tryCatch(withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error), error = handle_fatal,     skip = function(e) {    })
18: test_code(desc, code, env = parent.frame())
19: test_that("get_content works", {    scoped_experimental_silence()    content_list <- get_content(test_conn_1)    expect_is(content_list, c("tbl_df", "tbl", "data.frame"))    expect_identical(vctrs::vec_ptype(content_list), vctrs::vec_ptype(connectapi_ptypes$content))})
20: eval(code, test_env)
21: eval(code, test_env)
22: withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error)
23: doTryCatch(return(expr), name, parentenv, handler)
24: tryCatchOne(expr, names, parentenv, handlers[[1L]])
25: tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
26: doTryCatch(return(expr), name, parentenv, handler)
27: tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]),     names[nh], parentenv, handlers[[nh]])
28: tryCatchList(expr, classes, parentenv, handlers)
29: tryCatch(withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error), error = handle_fatal,     skip = function(e) {    })
30: test_code(NULL, exprs, env)
31: source_file(path, new.env(parent = env), chdir = TRUE, wrap = wrap)
32: force(code)
33: doWithOneRestart(return(expr), restart)
34: withOneRestart(expr, restarts[[1L]])
35: withRestarts(testthat_abort_reporter = function() NULL, force(code))
36: with_reporter(reporter = reporter, start_end_reporter = start_end_reporter,     {        reporter$start_file(basename(path))        lister$start_file(basename(path))        source_file(path, new.env(parent = env), chdir = TRUE,             wrap = wrap)        reporter$.end_context()        reporter$end_file()    })
37: FUN(X[[i]], ...)
38: lapply(paths, test_file, env = env, reporter = current_reporter,     start_end_reporter = FALSE, load_helpers = FALSE, wrap = wrap)
39: force(code)
40: doWithOneRestart(return(expr), restart)
41: withOneRestart(expr, restarts[[1L]])
42: withRestarts(testthat_abort_reporter = function() NULL, force(code))
43: with_reporter(reporter = current_reporter, results <- lapply(paths,     test_file, env = env, reporter = current_reporter, start_end_reporter = FALSE,     load_helpers = FALSE, wrap = wrap))
44: test_files(paths, reporter = reporter, env = env, stop_on_failure = stop_on_failure,     stop_on_warning = stop_on_warning, wrap = wrap)
45: test_dir("integrated-tests")
46: eval(code, test_env)
47: eval(code, test_env)
48: withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error)
49: doTryCatch(return(expr), name, parentenv, handler)
50: tryCatchOne(expr, names, parentenv, handlers[[1L]])
51: tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
52: doTryCatch(return(expr), name, parentenv, handler)
53: tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]),     names[nh], parentenv, handlers[[nh]])
54: tryCatchList(expr, classes, parentenv, handlers)
55: tryCatch(withCallingHandlers({    eval(code, test_env)    if (!handled && !is.null(test)) {        skip_empty()    }}, expectation = handle_expectation, skip = handle_skip, warning = handle_warning,     message = handle_message, error = handle_error), error = handle_fatal,     skip = function(e) {    })
56: test_code(NULL, exprs, env)
57: source_file(path, new.env(parent = env), chdir = TRUE, wrap = wrap)
58: force(code)
59: doWithOneRestart(return(expr), restart)
60: withOneRestart(expr, restarts[[1L]])
61: withRestarts(testthat_abort_reporter = function() NULL, force(code))
62: with_reporter(reporter = reporter, start_end_reporter = start_end_reporter,     {        reporter$start_file(basename(path))        lister$start_file(basename(path))        source_file(path, new.env(parent = env), chdir = TRUE,             wrap = wrap)        reporter$.end_context()        reporter$end_file()    })
63: FUN(X[[i]], ...)
64: lapply(paths, test_file, env = env, reporter = current_reporter,     start_end_reporter = FALSE, load_helpers = FALSE, wrap = wrap)
65: force(code)
66: doWithOneRestart(return(expr), restart)
67: withOneRestart(expr, restarts[[1L]])
68: withRestarts(testthat_abort_reporter = function() NULL, force(code))
69: with_reporter(reporter = current_reporter, results <- lapply(paths,     test_file, env = env, reporter = current_reporter, start_end_reporter = FALSE,     load_helpers = FALSE, wrap = wrap))
70: test_files(paths, reporter = reporter, env = env, stop_on_failure = stop_on_failure,     stop_on_warning = stop_on_warning, wrap = wrap)
71: (function (path, filter = NULL, reporter = default_reporter(),     env = test_env(), ..., encoding = "unknown", load_helpers = TRUE,     stop_on_failure = FALSE, stop_on_warning = FALSE, wrap = TRUE) {    if (!missing(encoding) && !identical(encoding, "UTF-8")) {        warning("`encoding` is deprecated; all files now assumed to be UTF-8",             call. = FALSE)    }    testthat_dir <- maybe_root_dir(path)    withr::local_envvar(list(R_TESTS = "", TESTTHAT = "true",         TESTTHAT_DIR = testthat_dir))    if (load_helpers) {        source_test_helpers(path, env)    }    source_test_setup(path, env)    on.exit(source_test_teardown(path, env), add = TRUE)    if (identical(Sys.getenv("NOT_CRAN"), "true")) {        withr::local_options(list(lifecycle_verbose_retirement = TRUE))    }    paths <- find_test_scripts(path, filter, ...)    test_files(paths, reporter = reporter, env = env, stop_on_failure = stop_on_failure,         stop_on_warning = stop_on_warning, wrap = wrap)})("/Users/carendt/rstudio/connectapi/tests/testthat", filter = NULL,     env = <environment>, stop_on_failure = FALSE, load_helpers = FALSE)
72: do.call(testthat::test_dir, testthat_args)
73: force(code)
74: withr::with_envvar(c(r_env_vars(), TESTTHAT_PKG = pkg$package),     do.call(testthat::test_dir, testthat_args))
75: force(code)
76: withr::with_options(c(useFancyQuotes = FALSE), withr::with_envvar(c(r_env_vars(),     TESTTHAT_PKG = pkg$package), do.call(testthat::test_dir,     testthat_args)))
77: force(code)
78: withr::with_collate("C", withr::with_options(c(useFancyQuotes = FALSE),     withr::with_envvar(c(r_env_vars(), TESTTHAT_PKG = pkg$package),         do.call(testthat::test_dir, testthat_args))))
79: devtools::test()
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault: 11
```

**Comment by colearendt** [comment]
Update: actually, `expect_identical` looks to be the culprit here. 

Using `expect_equal` instead gives this little gem:

```
R(33056,0x1149735c0) malloc: *** error for object 0x7facf2f12f10: pointer being freed was not allocated
R(33056,0x1149735c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

**Comment by hadley** [comment]
This is a dplyr issue that's already been fixed in dplyr 1.0.0, so I'm going to close this issue.
